Final net_pkt API switch

tbursztyka opened the pull request:
That will be the very last move to new API

Commit:
net/context: Re-enable offloaded send/sendto

net_offload API is left untouched, so it still takes a net_pkt as input
for the buffer. This is under-optimized since offload drivers will copy
the data from that net_pkt back into contiguous buffer again.
Let's tackle this issue another time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Remove legacy API

Former net_context_send/sendto and net_context_create_ipv4/ipv6 are now
unused and can be removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/udp: Remove net_udp_insert legacy function

It is unused anywhere now.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove appdata/appdatalen attributes and related functions

There is no need for these anymore: all is dictated by the position of
the net_pkt's cursor now

- actual cursor position is like the former appdata attribute
- net_pkt_remaining_data() is like the former appdatalen attribute

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Remove _new suffix on net_context_send/sendto functions

Now that legacy functions are removew, let's rename the new functions by
removing the _new suffix.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv4: Replace legacy net_ipv4_create by the new one

Thus removing the legacy one, and renaming the new one to legacy name.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Replace legacy net_ipv6_create by the new one

Thus removing the legacy one, and renaming the new one to legacy name.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Remove unused net_icmpv4_set_checksum function

Checksum is now set via net_icmpv4_finalize() function.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove now useless net_pkt_get_data()

That function was responsible for allocating new buffer element, but it
is now unused and can be removed safely. Buffer allocation is now done
via net_pkt_alloc_buffer().

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove _new suffix to net_pkt_get_data_new function

Now that legacy - and unrelated - function named net_pkt_get_data has
been removed, we can rename net_pkt_get_data_new relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove _new suffix to net_pkt_read functions

Suffix is now useless, as these functions are now the only ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove legacy net_pkt_write functions

These were based on the former allocation scheme and are no longer used.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove legacy net_pkt_insert functions

These were based on the former allocation scheme and are no longer used.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove _new suffix to net_pkt_write functions

Suffix is now useless, as these functions are now the only ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove legacy net_frag_read/skip/get_pos functions

These were based on the former allocation scheme and are no longer used.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove legacy net_pkt_append functions

These were based on the former allocation scheme and are no longer used.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove legacy net_frag_linear_copy function

This function is now superseded by net_pkt_read() and is no longer
used.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove unused token attribute

Seems like a useless attribute. Since net_context is not being used by
the user directly (socket is the unique interface now) and since no core
parts uses the token parameter of net_context API: let's remove the
attribute.

This helps to save 4 bytes from struct net_pkt.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Remove token parameter from net_context_send/sendto

And also to the relevant callbacks.

That parameter is not used anywhere so it is useless.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove superfluous total_pkt_len attribute

This attribute, in case CONFIG_NET_STATISTICS is enabled, made sense
when L2's send() function did not return the length of the sent packet.
But now, it's a superflous optimization as is it used only to set the
stats on recv or send, where net_pkt_get_len() can be used directly.

This helps to save 2 bytes from struct net_pkt.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
@tbursztyka, this needs rebase.

mike-scott commented:
@tbursztyka Do you have a time frame for when NET_OFFLOAD will be fixed in this series?

pfalcon commented:
> @tbursztyka Do you have a time frame for when NET_OFFLOAD will be fixed in this series?

Actually, are there any time frames and status updates on this PR at all? This has 48 commits, "WIP" status, and conflicts. And "we have 2 months left before the release" has now morphed into "we have 1 month left". So, any ETA, suggestions, guidelines are appreciated.

mike-scott commented:
@tbursztyka I can't really review/test w/ WNC-M14A2A modem until this PR fixes NET_OFFLOAD.  Is that done yet?

tbursztyka commented:
@mike-scott yes I have a local branch which I need to merge to this pr

Commit:
net/offload: Remove useless token parameter

This parameter was removed from net_context already thus applying the
change also on net_offload API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@mike-scott rebased against my offload branch. (I am away from my hw stash, so could not test it against winc1500 for instance)

Note that I still need to apply the changes to net/tests.

tbursztyka commented:
There are still one or two samples that new to switch to new net_pkt/net_context API, and 6lo test afaik.

mike-scott commented:
> @mike-scott rebased against my offload branch. (I am away from my hw stash, so could not test it against winc1500 for instance)

@tbursztyka I'll check this out into my shields WIP branch and kick the tires today.  Thanks for the update.

mike-scott commented:
@tbursztyka I tested this PR and NET_OFFLOAD is broken, but not entirely due to your rework.  This commit by @jukkar broke it initially: https://github.com/zephyrproject-rtos/zephyr/commit/fc27a81ed2c1b39c64c70502af3df2d17d7ff831#diff-0fca2d20b5935f1677a31c9a6e2b5087

The call to bind_default() was move till after the NET_OFFLOAD checks.  Inside bind_default() we call net_context_bind() which sets the net_if for the context.  This is the first error causing the abort I was seeing: a NULL iface passed into net_offload_connect().

The next error I saw was that context->remote handling was also being skipped due to the location of the NET_OFFLOAD handling.  During that handling we set context->flag |= NET_CONTEXT_REMOTE_ADDR_SET.  If that's not set when we call sendto() then we also get an error (EREQDESTADDR).

The fix was oddly simple, but I have a feeling there are probably side-effects (like for a listening socket, etc):  I moved the net_offload handling block down below the remote/local address handling.  Diff below:
```
diff --git a/subsys/net/ip/net_context.c b/subsys/net/ip/net_context.c
index 69cf07226b..9c36048dd8 100644
--- a/subsys/net/ip/net_context.c
+++ b/subsys/net/ip/net_context.c
@@ -916,20 +916,6 @@ int net_context_connect(struct net_context *context,
 		return -EOPNOTSUPP;
 	}
 
-#if defined(CONFIG_NET_OFFLOAD)
-	if (net_if_is_ip_offloaded(net_context_get_iface(context))) {
-		ret = net_offload_connect(
-			net_context_get_iface(context),
-			context,
-			addr,
-			addrlen,
-			cb,
-			timeout,
-			user_data);
-		goto unlock;
-	}
-#endif /* CONFIG_NET_OFFLOAD */
-
 	if (net_context_get_state(context) == NET_CONTEXT_LISTENING) {
 		ret = -EOPNOTSUPP;
 		goto unlock;
@@ -1042,6 +1028,20 @@ int net_context_connect(struct net_context *context,
 		goto unlock;
 	}
 
+#if defined(CONFIG_NET_OFFLOAD)
+	if (net_if_is_ip_offloaded(net_context_get_iface(context))) {
+		ret = net_offload_connect(
+			net_context_get_iface(context),
+			context,
+			addr,
+			addrlen,
+			cb,
+			timeout,
+			user_data);
+		goto unlock;
+	}
+#endif /* CONFIG_NET_OFFLOAD */
+
 	switch (net_context_get_type(context)) {
 #if defined(CONFIG_NET_UDP)
 	case SOCK_DGRAM:
```

tbursztyka commented:
@mike-scott Thanks for trying this, actually you could just have complained it's buggy :) I appreciate the investigations! I'll get back to that as soonish.

mike-scott committed:
net/context: Relocate offload handling in net_context_connect()

Commit fc27a81ed2c1 ("net: context: Select proper network interface
when binding") moved the bind_default() call to after the remote
address was set for the net_context.

This and a later net_pkt API refactor broke net_offload() handling
so that context->iface wasn't set and context->flags didn't have
NET_CONTEXT_REMOTE_ADDR_SET correctly.

Let's fix this by relocating the net_offload handling to after
these have happened.

Signed-off-by: Michael Scott <mike@foundries.io>

tbursztyka commented:
@mike-scott I re-reviewed the net_offload (still not trying it, I will be able to do that only tomorrow) and I fully agree with you conclusions. You patch is a valid fix to an existing issue, and not related to re-enabling offload. Could you make commit (with your signed-off-by) out of your patch that I will include in this PR? (I don't think I can do that myself or do I? Just putting your signed-off-by would be sufficient to keep your authorship on the patch?).

jhedberg commented:
Not sure I'll have the chance to do a thorough review, but I do like the look of the diffstat `+532 −2,655` :)

mike-scott commented:
> @mike-scott I re-reviewed the net_offload (still not trying it, I will be able to do that only tomorrow) and I fully agree with you conclusions. You patch is a valid fix to an existing issue, and not related to re-enabling offload. Could you make commit (with your signed-off-by) out of your patch that I will include in this PR? (I don't think I can do that myself or do I? Just putting your signed-off-by would be sufficient to keep your authorship on the patch?).

@tbursztyka Feel free to pick up the patch from here: https://github.com/mike-scott/zephyr/commits/final_switch

Commit:
net: Switch usage of net_pkt_get_reserve to net_pkt_alloc

Some places were still using the old allocator. Using the new one does
not change any behavior. This will help to remove the useless data_len
attribute in net_pkt which legacy allocator was still setting.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove unused legacy net_pkt allocator

Now, only net_pkt_alloc and variants are used.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Remove now useless attribute data_len

This reduces net_pkt structure size.
Moving sent_list attribute for better alignment.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@mike-scott thanks, commit added.

@jhedberg and there is room for more removal actually

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -371,10 +367,8 @@ static void net_queue_rx(struct net_if *iface, struct net_pkt *pkt)
-
 	net_stats_update_tc_recv_pkt(iface, tc);
-	net_stats_update_tc_recv_bytes(iface, tc, pkt->total_pkt_len);
+	net_stats_update_tc_recv_bytes(iface, tc, net_pkt_get_len(pkt));
```
We need to figure out better way to do this at some point. The `net_pkt_get_len()` will traverse the buffer list and sum the total length. This is now done multiple times in worst case if CONFIG_NET_STATISTICS is enabled, first time in `net_queue_rx()` and then again in `net_rx()`

tbursztyka replied:
Actually I did have a way to improve that in my prototype back then, basically having a length attribute in net_pkt, that was affected by net_pkt_writ/skip/memset/pull if overwrite was false.
But I wonder if that's necessary, on ethernet frames if we take a net_buf size of 128 bytes we would have at worst 11 net_buf. Of course if we call net_pkt_get_len() a lot, then that's another issue. It's a trade off I guess, whether we want to save 2bytes from net_pkt of some cpu. A bit of benchmarking wouldn't hurt to figure out, let's see this in a near future.

jukkar reviewed (approved):
LGTM

rveerama1 reviewed (approved):
LGTM

mike-scott commented on drivers/modem/wncm14a2a.c:
```diff
@@ -464,7 +464,7 @@ static int pkt_setup_ip_data(struct net_pkt *pkt,
-		if (net_ipv6_create_new(
+		if (net_ipv6_create(
 			    pkt,
 			    &((struct sockaddr_in6 *)&sock->src)->sin6_addr,
```
Oh, I meant to also point out the net_ipv6_create params (src and dst) are flipped here.  Vs the net_ipv4_create() call right below.  Was in your previous patch series which has already been merged.  (Found it while testing)

tbursztyka replied:
Go ahead, get a dedicated PR for it as it is a bug fix so it can be applied asap (this current PR has tsc label so it will wait 3 days still)

mike-scott replied:
Will do.

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1265,7 +1265,8 @@ static int context_sendto(struct net_context *context,
 			  s32_t timeout,
-			  void *user_data)
+			  void *user_data,
+			  bool sendto)
```
This should not be needed, as send() is strictly equivalent to sendto(dst_addr=NULL). So, the right solution would be to drop net_offload_send().

tbursztyka replied:
There is definitely things to improve on net_offload and its integration within net_context but it's too late for now. So let's keep such ideas for 1.15

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1265,7 +1265,8 @@ static int context_sendto(struct net_context *context,
 			  socklen_t addrlen,
```
Commit message typo:

> net/context: Re-enable offloaded send/sendto 
> net_offload API is let untouched

Apparently, "left untouched".

tbursztyka replied:
yep

pfalcon commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1120,7 +1120,7 @@ static struct net_pkt *frame_get(struct gmac_queue *queue)
 		return NULL;
```
Commit message typo:

> net: Switch usage of net_pkt_get_reserve to net_pkt_alloc 
> Some places where still using the old allocator.

Apparently "were using".

tbursztyka replied:
yep

mike-scott commented:
I've been testing this for a while, and with @pfalcon comments address, it LGTM.

rerickson1 commented:
I can also confirm that NET_OFFLOAD has been working for me (with 3872cba) during my new modem driver development.

pfalcon reviewed (approved):
+1 too for soonish merging
